compiler/rustc_middle/src/ty/print/pretty.rs

```diff
@@ -1281,19 +1281,48 @@ pub trait PrettyPrinter<'tcx>: Printer<'tcx> + fmt::Write {
                 // HACK(eddyb) this duplicates `FmtPrinter`'s `path_generic_args`,
                 // in order to place the projections inside the `<...>`.
                 if !resugared {
-                    // Use a type that can't appear in defaults of type parameters.
-                    let dummy_cx = Ty::new_fresh(cx.tcx(), 0);
-                    let principal = principal.with_self_ty(cx.tcx(), dummy_cx);
+                    let principal_with_self =
+                        principal.with_self_ty(cx.tcx(), cx.tcx().types.trait_object_dummy_self);
 
                     let args = cx
                         .tcx()
-                        .generics_of(principal.def_id)
-                        .own_args_no_defaults(cx.tcx(), principal.args);
+                        .generics_of(principal_with_self.def_id)
+                        .own_args_no_defaults(cx.tcx(), principal_with_self.args);
+
+                    let bound_principal_with_self = bound_principal
+                        .with_self_ty(cx.tcx(), cx.tcx().types.trait_object_dummy_self);
+
+                    let super_projections: Vec<_> =
+                        super_predicates_for_pretty_printing(cx.tcx(), bound_principal_with_self)
+                            .filter_map(|clause| clause.as_projection_clause())
+                            .collect();
+
+                    let mut projections: Vec<_> = predicates
+                        .projection_bounds()
+                        .filter_map(|proj| {
+                            // Filter out projections that are implied by the super predicates.
+                            let proj_is_implied = super_projections.iter().any(|&super_proj| {
+                                let proj = cx.tcx().anonymize_bound_vars(proj);
+                                let super_proj = cx.tcx().anonymize_bound_vars(super_proj);
+                                assert_eq!(proj.bound_vars(), super_proj.bound_vars());
```

Comment:
This isn't true:

```
#![feature(generic_associated_types_extended)]

struct A;
struct B;

trait Foo: for<'a> Bar<'a, A, Output = ()> + Bar<'static, B, Output = ()> {}

trait Bar<'t, T> {
    type Output;
}

fn foo(mut x: &dyn Foo) {
    x = ();
}

fn main() {}
```

This code ICEs.

Please remove the assertion.

Reply:
I am also skeptical that the anonymization is necessary here -- do you find that it actually affects? Correct super-predicate substitution should mean that it doesn't really matter here.

Reply:
The anonymization is necessary, because this function is sometimes called on types with anonymized bound regions, for example for the expected/found message [here](https://github.com/rust-lang/rust/pull/120739/commits/18e5bbfad6c6905704ac111f9fdffbbe638c8e33#diff-07495ab614c7036a558aea5c0c23d7c94cf288e4057e88c0708e3bfe9165e471L115). But `tcx.super_predicates_of` will always return the named regions and after super-predicate substitution we end up with a type that has a mix of named regions and anonymized regions. In that case comparing the fully anonymized `proj` and half-anonymized `super_proj` will give false, even though they are equal modulo region names.

I will add a comment.